src/main/java/org/dependencytrack/notification/ScheduledNotificationTaskManager.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * This file is part of Dependency-Track.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright (c) OWASP Foundation. All Rights Reserved.
+ */
+package org.dependencytrack.notification;
+
+import java.util.HashMap;
+import java.util.UUID;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.dependencytrack.tasks.ActionOnDoneFutureTask;
+import org.dependencytrack.tasks.SendScheduledNotificationTask;
+
+import com.asahaf.javacron.Schedule;
+
+/* 
+ * Holds all scheduled notification tasks to be able to cancel them if needed.
+ * This may be necessary if a rule is deleted or disabled.
+ */
+public final class ScheduledNotificationTaskManager {
+    private static final HashMap<UUID, ScheduledFuture<?>> SCHEDULED_NOTIFY_TASKS = new HashMap<UUID, ScheduledFuture<?>>();
+
+    public static void scheduleNextRuleTask(UUID ruleUuid, Schedule schedule, long customDelay, TimeUnit delayUnit) {
+        scheduleNextRuleTask(ruleUuid, schedule, customDelay, delayUnit, () -> scheduleNextRuleTask(ruleUuid, schedule));
+    }
+
+    public static void scheduleNextRuleTask(UUID ruleUuid, Schedule schedule) {
+        scheduleNextRuleTask(ruleUuid, schedule, schedule.nextDuration(TimeUnit.MILLISECONDS), TimeUnit.MILLISECONDS);
+    }
+
+    public static void scheduleNextRuleTaskOnce(UUID ruleUuid, long customDelay, TimeUnit delayUnit){
+        scheduleNextRuleTask(ruleUuid, null, customDelay, delayUnit, () -> cancelActiveRuleTask(ruleUuid));
+    }
+
+    private static void scheduleNextRuleTask(UUID ruleUuid, Schedule schedule, long customDelay, TimeUnit delayUnit, Runnable actionAfterTaskCompletion){
+        var scheduledExecutor = Executors.newSingleThreadScheduledExecutor();
+        var futureTask = new ActionOnDoneFutureTask(new SendScheduledNotificationTask(ruleUuid), actionAfterTaskCompletion);
```

Comment:
You can use `CallbackEvent`s to execute arbitrary code when processing of a given `Event` completed successfully. You can see it being used here to decrement a `CountdownLatch`:

https://github.com/DependencyTrack/dependency-track/blob/af4461e86ae56eaee193017ec86ee2a690709932/src/main/java/org/dependencytrack/tasks/metrics/PortfolioMetricsUpdateTask.java#L78-L86

Reply:
I looked a bit into that and i think using the Callback Event is not benifitial here. We need the cron-scheduled execution, so the runnable task for the scheduled executor with be needed nonetheless. I could rewrite it in a way, that the runnable will just dispatch the event and the subscriber will react to it by executing the SendScheduleNotificationTask (or scheduling a new one), but it's just another layer on the same logic. Please explain the benefit your seeing in this shortly, if you would like to change it that way.